File: CS/PropertyGridWithDictionary/CustomPropertyDescriptor.cs
using System;
using System.ComponentModel;

namespace PropertyGridWithDictionary {
    public class CustomPropertyDescriptor<T> : PropertyDescriptor {
        private Type _propertyType;

        private Type _componentType;

        private T _oldValue;

        public CustomPropertyDescriptor(Type componentType, string propertyName, Type propertyType, Attribute[] attributes)
            : base(propertyName, attributes) {
            _propertyType = propertyType;
            _componentType = componentType;
        }

        public override bool CanResetValue(object component) {
            if (_oldValue == null)
                return false;
            else
                return true;
        }

        public override Type ComponentType {
            get { return _componentType; }
        }

        public override object GetValue(object component) {
            return (component as DictionaryWrapper<T>).Dict[base.Name];
        }

        public override bool IsReadOnly {
            get { return false; }
        }

        public override Type PropertyType {
            get { return _propertyType; }
        }

        public override void ResetValue(object component) {
            SetValue(component, _oldValue);
        }

        public override void SetValue(object component, object value) {
            _oldValue = (component as DictionaryWrapper<T>).Dict[base.Name];
            (component as DictionaryWrapper<T>).Dict[base.Name] = (T)value;
        }

        public override bool ShouldSerializeValue(object component) {
            return false;
        }
    }
}

File: VB/PropertyGridWithDictionary/CustomPropertyDescriptor.vb
Imports System
Imports System.ComponentModel

Namespace PropertyGridWithDictionary

    Public Class CustomPropertyDescriptor(Of T)
        Inherits PropertyDescriptor

        Private _propertyType As Type

        Private _componentType As Type

        Private _oldValue As T

        Public Sub New(ByVal componentType As Type, ByVal propertyName As String, ByVal propertyType As Type, ByVal attributes As Attribute())
            MyBase.New(propertyName, attributes)
            _propertyType = propertyType
            _componentType = componentType
        End Sub

        Public Overrides Function CanResetValue(ByVal component As Object) As Boolean
            If _oldValue Is Nothing Then
                Return False
            Else
                Return True
            End If
        End Function

        Public Overrides ReadOnly Property ComponentType As Type
            Get
                Return _componentType
            End Get
        End Property

        Public Overrides Function GetValue(ByVal component As Object) As Object
            Return TryCast(component, DictionaryWrapper(Of T)).Dict(Name)
        End Function

        Public Overrides ReadOnly Property IsReadOnly As Boolean
            Get
                Return False
            End Get
        End Property

        Public Overrides ReadOnly Property PropertyType As Type
            Get
                Return _propertyType
            End Get
        End Property

        Public Overrides Sub ResetValue(ByVal component As Object)
            SetValue(component, _oldValue)
        End Sub

        Public Overrides Sub SetValue(ByVal component As Object, ByVal value As Object)
            _oldValue = TryCast(component, DictionaryWrapper(Of T)).Dict(Name)
            TryCast(component, DictionaryWrapper(Of T)).Dict(Name) = CType(value, T)
        End Sub

        Public Overrides Function ShouldSerializeValue(ByVal component As Object) As Boolean
            Return False
        End Function
    End Class
End Namespace
